src/EditorFeatures/Core/Shared/Tagging/Utilities/TagSpanIntervalTree.cs

```diff
@@ -98,51 +110,53 @@ private static void DebugVerifyTags(NormalizedSnapshotSpanCollection requestedSp
+        private void AddIntersectingTagSpansWorker(
+            NormalizedSnapshotSpanCollection requestedSpans,
+            SegmentedList<ITagSpan<TTag>> tags)
         {
             const int MaxNumberOfRequestedSpans = 100;
 
             // Special case the case where there is only one requested span.  In that case, we don't
             // need to allocate any intermediate collections
-            return requestedSpans.Count == 1
-                ? GetIntersectingSpans(requestedSpans[0])
-                : requestedSpans.Count < MaxNumberOfRequestedSpans
-                    ? GetTagsForSmallNumberOfSpans(requestedSpans)
-                    : GetTagsForLargeNumberOfSpans(requestedSpans);
+            if (requestedSpans.Count == 1)
+                AddIntersectingSpans(requestedSpans[0], tags);
+            else if (requestedSpans.Count < MaxNumberOfRequestedSpans)
+                AddTagsForSmallNumberOfSpans(requestedSpans, tags);
+            else
+                AddTagsForLargeNumberOfSpans(requestedSpans, tags);
         }
 
-        private IEnumerable<ITagSpan<TTag>> GetTagsForSmallNumberOfSpans(NormalizedSnapshotSpanCollection requestedSpans)
+        private void AddTagsForSmallNumberOfSpans(
+            NormalizedSnapshotSpanCollection requestedSpans,
+            SegmentedList<ITagSpan<TTag>> tags)
         {
-            var result = new List<ITagSpan<TTag>>();
-
-            foreach (var s in requestedSpans)
-            {
-                result.AddRange(GetIntersectingSpans(s));
-            }
-
-            return result;
+            foreach (var span in requestedSpans)
+                AddIntersectingSpans(span, tags);
         }
 
-        private IEnumerable<ITagSpan<TTag>> GetTagsForLargeNumberOfSpans(NormalizedSnapshotSpanCollection requestedSpans)
+        private void AddTagsForLargeNumberOfSpans(NormalizedSnapshotSpanCollection requestedSpans, SegmentedList<ITagSpan<TTag>> tags)
         {
             // we are asked with bunch of spans. rather than asking same question again and again, ask once with big span
             // which will return superset of what we want. and then filter them out in O(m+n) cost. 
             // m == number of requested spans, n = number of returned spans
-            var mergedSpan = new SnapshotSpan(requestedSpans[0].Start, requestedSpans[requestedSpans.Count - 1].End);
-            var result = GetIntersectingSpans(mergedSpan);
+            var mergedSpan = new SnapshotSpan(requestedSpans[0].Start, requestedSpans[^1].End);
 
-            var requestIndex = 0;
+            using var _1 = Classifier.GetPooledList<ITagSpan<TTag>>(out var tempList);
 
-            var enumerator = result.GetEnumerator();
+            AddIntersectingSpans(mergedSpan, tempList);
+            if (tempList.Count == 0)
+                return;
+
+            var requestIndex = 0;
+            var enumerator = tempList.GetEnumerator();
```

Comment:
>var enumerator = tempList.GetEnumerator(); [](http://example.com/codeflow?start=12&length=42)

should there be a using on this so that the code below doesn't need to worry about disposal?

Reply:
Yup. I can make that change

Reply:
done.